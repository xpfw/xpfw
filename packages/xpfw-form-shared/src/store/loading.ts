import { IFieldError, IFormError } from "@xpfw/validate"
import { get, set } from "lodash"
import { autoSubscribe, AutoSubscribeStore, key, StoreBase } from "resub"
import { IPersistableStore } from "resub-persist"

@AutoSubscribeStore
export class LoadingStore extends StoreBase implements IPersistableStore {
    public name = "loading"
    private loading: {[path: string]: boolean} = {}

    public getPropKeys() { return ["loading"] }

    public setLoading(valuePath: string, value: any) {
      if (value === true) {
        this.loading[valuePath] = true
      } else {
        delete this.loading[valuePath]
      }
      this.trigger(valuePath)
    }

    @autoSubscribe
    public getLoading(@key path: string): boolean {
      return this.loading[path] ? true : false
    }

    @autoSubscribe
    public isSomethingLoading(): boolean {
      return Object.keys(this.loading).length > 0
    }
}

export default new LoadingStore()
