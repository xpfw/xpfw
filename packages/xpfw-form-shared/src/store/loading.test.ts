import { matchStoreState } from "resub-persist"
import { LoadingStore } from "./loading"

const LoadingStor = new LoadingStore()
const path = "lp"

test("Cover Form Store", () => {
  matchStoreState(LoadingStor, "Initial State")
  expect(LoadingStor.getLoading(path)).toBeFalsy()
  LoadingStor.setLoading(path, true)
  expect(LoadingStor.getLoading(path)).toBeTruthy()
  LoadingStor.setLoading(path + path, true)
  matchStoreState(LoadingStor, "two things loading")
  LoadingStor.setLoading(path, "true") // anything that isn't === true results in false / deletion
  matchStoreState(LoadingStor, "one thing loading")
  expect(LoadingStor.isSomethingLoading()).toBeTruthy()
  LoadingStor.setLoading(path + path, false)
  LoadingStor.setLoading(path, "nothing loading anymore")
})
