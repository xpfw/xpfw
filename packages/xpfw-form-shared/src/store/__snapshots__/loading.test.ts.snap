// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cover Form Store: Initial State 1`] = `
Object {
  "loading": Object {},
}
`;

exports[`Cover Form Store: one thing loading 1`] = `
Object {
  "loading": Object {
    "lplp": true,
  },
}
`;

exports[`Cover Form Store: two things loading 1`] = `
Object {
  "loading": Object {
    "lp": true,
    "lplp": true,
  },
}
`;
